locator/abstract_replication_strategy.cc

```diff
@@ -178,9 +184,11 @@ abstract_replication_strategy::get_ranges(inet_address ep, token_metadata_ptr tm
     if (sorted_tokens.empty()) {
         on_internal_error(rslogger, "Token metadata is empty");
     }
+    std::optional<inet_address_vector_replica_set> eps;
     auto prev_tok = sorted_tokens.back();
     for (auto tok : sorted_tokens) {
-        for (inet_address a : co_await calculate_natural_endpoints(tok, tm)) {
+        co_await maybe_calculate_natural_endpoints(eps, tok, tm);
```

Comment:
I think that for everywhere and local, ignoring vnodes will yield a much bigger benefit (especially everywhere which is quadratic with vnodes).

Reply:
I tend to agree but there are edge cases where we currently return the local endpoint while token_metadata is empty and after it is populated we return all nodes.  To do that we'll probably need to override several abstract_replication_strategy methods in everywhere r.s. and local strategy so they'll have their simplified implementation.

Maybe I'll get back to it when working on compaction groups.

Reply:
Can you elaborate? Why is token_metadata empty?

Reply:
IIRC when bootstrapping a node token_metadata is initially empty until gossip decides about the bootstrap tokens

Reply:
Another point of concern is that the token_metadata is currently shared to all replication strategies and there is a single (versioned) source of truth for it, and it is sampled by the effective_,replication_map.  For having a single token for local/everywhere strategies we'll need to decouple their token_metadata from the rest.  It's probably possible, but will require delicate surgery.